[Storage] `AppendBlobClient`, `PageBlobClient`, and initial `create` APIs

vincenttran-msft opened the pull request:
WIP

vincenttran-msft commented on sdk/storage/azure_storage_blob/tests/page_blob_client.rs:
```diff
@@ -0,0 +1,143 @@
+        .await?;
+
+    page_blob_client
+        .clear_page("bytes=0-511".to_string(), None)
```
Same here for `clear_page`

vincenttran-msft commented on sdk/storage/azure_storage_blob/tests/page_blob_client.rs:
```diff
@@ -0,0 +1,164 @@
+    assert_eq!(BlobType::PageBlob, blob_type.unwrap());
+
+    // Create If Not Exists Scenario
+    let create_options = PageBlobClientCreateOptions {
+        if_none_match: Some("*".to_string()),
+        ..Default::default()
```
Wondering if there is any cool ways we could make a helper for this i.e. like somehow have a `.if_not_exists()` that you can call on the options bag -- another pattern we haven't yet discussed (how to go about modifying customer's options bag)

vincenttran-msft commented on sdk/storage/azure_storage_blob/tests/page_blob_client.rs:
```diff
@@ -28,10 +29,7 @@ async fn test_create_page_blob(ctx: TestContext) -> Result<(), Box<dyn Error>> {
-        if_none_match: Some("*".to_string()),
-        ..Default::default()
-    };
+    let create_options = page_blob_create_if_not_exists(PageBlobClientCreateOptions::default());
```
Wondering if there is a better way to achieve this effect? We would like to abstract out the logic to get this to be "create if not exists". 

Even better if we could make this generic, as this is likely not the only API that would support "if not exists". Maybe a Trait of some sort and we somehow mark all the applicable option bags with this Trait?

vincenttran-msft replied:
Closed as duplicate on megathread on helpers.

jalauzon-msft commented on sdk/storage/azure_storage_blob/src/clients/page_blob_client.rs:
```diff
@@ -0,0 +1,166 @@
+    /// * `size` - Size used to resize the blob. Maximum size for a page Blob is up to 1TB. The
+    ///   Page blob size must be aligned to a 512-byte boundary.
+    /// * `options` - Optional parameters for the request.
+    pub async fn resize_blob(
```
If we are sticking with past patterns, this should probably just be `resize`?

heaths commented on sdk/storage/azure_storage_blob/src/serialize.rs:
```diff
@@ -0,0 +1,31 @@
+///   The offset specified must be a modulus of 512.
+/// * `length` - Number of bytes to use for writing to a section of the blob.
+///   The length specified must be a modulus of 512.
+pub fn format_http_range(offset: u64, length: u64) -> String {
+    if offset % 512 != 0 {
+        panic!("offset must be aligned to a 512-byte boundary.");
+    }
+    if length % 512 != 0 {
+        panic!("length must be aligned to a 512-byte boundary.");
+    }
+    let end_range = offset + length - 1;
+    let content_range = format!("bytes={}-{}", offset, end_range);
+    content_range
+}
+
+pub fn page_blob_create_if_not_exists(
+    mut options: PageBlobClientCreateOptions,
+) -> PageBlobClientCreateOptions {
+    options.if_none_match = Some("*".to_string());
+    options
+}
```
Why an unassociated function? These aren't as discoverable. You're also taking a `mut` by value, which is odd since that puts a burden on callers and you're just returning a "copy" (moved, actually) anyway. Use the `Self { if_none_match = Some("*".into()), ..self }` syntax instead so you don't need `mut`, which ends up doing the same thing anyway.

But if you want to repeat the pattern, take a page from the convenience client guidelines: https://azure.github.io/azure-sdk/rust_implementation.html#rust-client-convenience

Define a trait that you can implement for different method option types. It also means you don't need unassociated functions like this.

heaths commented on sdk/storage/azure_storage_blob/src/lib.rs:
```diff
@@ -29,7 +31,12 @@ pub mod models {
+        PageBlobClientUploadPagesResult, PublicAccessType, RehydratePriority,
+        StorageServiceProperties,
     };
+    pub use crate::serialize::*;
```
See https://azure.github.io/azure-sdk/rust_implementation.html#rust-client-convenience for guidelines. In general, helpers should be in the highest-level module for the types they augment and end with "Ext".

vincenttran-msft commented on sdk/storage/azure_storage_blob/src/storage_traits.rs:
Based on the feedback, I moved both helpers to a `storage_traits.rs` file. Let me know if any further changes should be made 😄

heaths commented on sdk/storage/azure_storage_blob/tests/page_blob_client.rs:
```diff
@@ -29,7 +28,47 @@ async fn test_create_page_blob(ctx: TestContext) -> Result<(), Box<dyn Error>> {
 
     // Create If Not Exists Scenario
-    let create_options = page_blob_create_if_not_exists(PageBlobClientCreateOptions::default());
+    let create_options = PageBlobClientCreateOptions::default().with_if_not_exists();
```
I was thinking more about this last night, and the extension method idea might not work well here for client method options. It seems like what you really want is a custom client method options, in which case you could define your own that converts into a generated one and never expose the generated one, much like I did in #2737 recently e.g.,

```rust
use crate::{azure_core::http::ClientMethodOptions, generated::models::Tag};
use std::collections::HashMap;

#[derive(Debug, Default)]
pub struct UploadOptions {
    pub method_options: ClientMethodOptions,
    pub tags: HashMap<String, String>,
}

impl From<UploadOptions> for generated::models::UploadOptions {
    fn from(value: UploadOptions) -> Self {
        Self {
            method_options: value.method_options,
            tags: value
                .tags
                .into_iter()
                .map(|(k, v)| Tag {
                    key: Some(k.into()),
                    value: Some(v.into()),
                })
                .collect(),
        }
    }
}

fn main() {
    let mut options = UploadOptions::default();
    options.tags.insert("foo".into(), "1".into());
    options.tags.insert("bar".into(), "baz".into());

    upload(Some(options));
}

fn upload(options: Option<UploadOptions>) {
    let options: generated::models::UploadOptions = options.unwrap_or_default().into();
    println!("{options:#?}");
}

mod generated {
    pub mod models {
        use crate::azure_core::http::ClientMethodOptions;

        #[derive(Debug, Default)]
        pub struct Tag {
            pub key: Option<String>,
            pub value: Option<String>,
        }

        #[derive(Debug, Default)]
        pub struct UploadOptions {
            pub method_options: ClientMethodOptions,
            pub tags: Vec<Tag>,
        }
    }
}

pub mod azure_core {
    pub mod http {
        #[derive(Debug, Default)]
        pub struct ClientMethodOptions;
    }
}
```

Because this is repeated for several client options, you could define a helper or even implement `From<HashMap<String, String>> into Tag` or something to avoid duplication.

vincenttran-msft commented on sdk/storage/azure_storage_blob/src/storage_traits.rs:
```diff
@@ -0,0 +1,21 @@
+    fn with_if_not_exists(self) -> Self;
+}
+
+impl PageBlobClientCreateOptionsExt for PageBlobClientCreateOptions<'_> {
+    fn with_if_not_exists(self) -> Self {
+        Self {
+            if_none_match: Some("*".into()),
+            ..self
+        }
+    }
+}
```
This is where we landed for the scenario: Options bag option mutation
We went with an extension method on the generated options bag

vincenttran-msft commented on sdk/storage/azure_storage_blob/src/storage_utils.rs:
```diff
@@ -0,0 +1,26 @@
+    ///   The offset specified must be a modulus of 512.
+    /// * `length` - Number of bytes to use for writing to a section of the blob.
+    ///   The length specified must be a modulus of 512.
+    pub fn format_http_range(offset: u64, length: u64) -> String {
+        if offset % 512 != 0 {
+            panic!("offset must be aligned to a 512-byte boundary.");
+        }
+        if length % 512 != 0 {
+            panic!("length must be aligned to a 512-byte boundary.");
+        }
+        let end_range = offset + length - 1;
+        let content_range = format!("bytes={}-{}", offset, end_range);
+        content_range
+    }
```
This is where we landed for the scenario: General use helper functions
Our current plan is that any of these general use will be implemented here and called out in the docstring.

vincenttran-msft commented on sdk/storage/azure_storage_blob/src/clients/page_blob_client.rs:
```diff
@@ -0,0 +1,158 @@
+    ///
+    /// * `content_length` - The maximum size for the Page blob, up to 1TB. The page blob size must
+    ///   be aligned to a 512-byte boundary.
+    /// * `options` - Optional parameters for the request. See [`azure_storage_blob::storage_traits::PageBlobClientCreateOptionsExt`](crate::storage_traits::PageBlobClientCreateOptionsExt) for additional usage helpers.
```
This is how we plan to ensure discoverability. Docs render:
![image](https://github.com/user-attachments/assets/43bca67a-1a4b-4286-8fb5-f5a40c862276)

heaths replied:
Don't make the whole path a link. For one, if it's already in scope, you don't need to qualify it. If not, use markdown to shorten it:
```suggestion
    /// * `options` - Optional parameters for the request. See [`PageBlobClientCreateOptionsExt`](crate::storage_traits::PageBlobClientCreateOptionsExt) for additional usage helpers.
```

vincenttran-msft commented on sdk/storage/azure_storage_blob/src/clients/page_blob_client.rs:
```diff
@@ -0,0 +1,158 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `range` - The range of bytes to clear. See [`azure_storage_blob::StorageUtils::format_http_range()`](crate::StorageUtils::format_http_range) for help with the expected String format.
```
This is how we plan to ensure discoverability of the unassociated functions. Docs render:
![image](https://github.com/user-attachments/assets/d2999b20-7975-4dee-9fb9-f8e10e0cbf42)

heaths replied:
Again, don't fully-qualify the whole thing. It's harder to read and unnecessary. It's a link, and clicking on it tells them all they need to know. The rust-analyzer LSP will import/qualify things as needed.

heaths commented on sdk/storage/azure_storage_blob/src/storage_utils.rs:
```diff
@@ -0,0 +1,26 @@
+
+pub struct StorageUtils;
+
+impl StorageUtils {
```
This isn't C#. ;) You can declare just functions. `azure_core` has quite a few, as do most of the crates we call. I would just define this as a standalone function. There's no benefit of putting it in a type that otherwise has no purpose.

heaths reviewed (changes requested):
Some refactoring requested, but otherwise LGTM

vincenttran-msft commented:
Latest revision pushed changes:

- Addressed PR feedback
- Moved `models` module to it's own directory + `mod.rs`
- Added `extensions.rs` (formerly `storage_traits.rs`) to `models` module that will hold all option bag extension methods
- Renamed `storage_utils.rs` to `helpers.rs` to hold all unassociated functions

heaths commented on sdk/storage/azure_storage_blob/src/lib.rs:
```diff
@@ -9,34 +9,8 @@
 
 pub mod clients;
 mod generated;
+pub mod helpers;
```
The module shouldn't be public, but just export it from the root: https://azure.github.io/azure-sdk/rust_implementation.html#rust-modules

Lots of little modules just makes things hard to find.

heaths replied:
And if it's not meant to be public, it should be `pub(crate)`: https://azure.github.io/azure-sdk/rust_introduction.html#rust-namespace-naming-internals

jalauzon-msft replied:
Hi @heaths, we did consider putting this at the root but felt it was a little better to not have these helpers pollute the root namespace. It just felt a little odd to do `azure_storage_blob::format_page_range` directly. We made the generic enough, so we don't add a bunch more modules later.

If it is truly more idiomatic to export these types of helpers from the root, so be it, just thought I would share my opinion.

heaths reviewed (changes requested):
I don't think we should expose a `helpers` module. What purpose does it serve? Seems better to expose it from the root if it applies pretty much crate-wide. That would follow guidelines.

vincenttran-msft commented:
@heaths Heath, We have decided to move from `helpers.rs` to `parsers.rs` and re-exported the module from the root. 

I am currently hitting a compilation error from `typespec_client_core` [here](https://dev.azure.com/azure-sdk/public/_build/results?buildId=5056802&view=logs&j=2c0cb571-58e4-5f13-9d77-617bed7e8990&t=3768c134-7568-558c-110f-6089b91d12de&l=405) even after merging in `main`, but other than that the PR is ready to review, thanks!
